cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,357 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/hashing.hpp>
+#include <cudf/detail/utilities/hash_functions.cuh>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/table/table_device_view.cuh>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+// SHA supported leaf data type check
+bool sha_type_check(data_type dt)
+{
+  return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
+}
+
+}  // namespace
+
+struct SHA1Hash {
+  CUDA_DEVICE_CALLABLE uint32_t rotl32(uint32_t x, int8_t r) const
+  {
+    // Equivalent to (x << r) | (x >> (32 - r))
+    return __funnelshift_l(x, x, r);
+  }
+
+  /**
+   * @brief Core SHA-1 algorithm implementation. Processes a single 512-bit chunk,
+   * updating the hash value so far. Does not zero out the buffer contents.
+   */
+  void __device__ hash_step(sha1_intermediate_data* hash_state) const
+  {
+    uint32_t A = hash_state->hash_value[0];
+    uint32_t B = hash_state->hash_value[1];
+    uint32_t C = hash_state->hash_value[2];
+    uint32_t D = hash_state->hash_value[3];
+    uint32_t E = hash_state->hash_value[4];
+
+    uint32_t words[80];
+
+    // The 512-bit message buffer fills the first 16 words.
+    for (int i = 0; i < 16; i++) {
+      uint32_t buffer_element_as_int;
+      std::memcpy(&buffer_element_as_int, hash_state->buffer + (i * 4), 4);
+      // Convert word representation from little-endian to big-endian.
+      words[i] = __byte_perm(buffer_element_as_int, 0, 0x0123);
+    }
+
+    // The rest of the 80 words are generated from the first 16 words.
+    for (int i = 16; i < 80; i++) {
+      uint32_t temp = words[i - 3] ^ words[i - 8] ^ words[i - 14] ^ words[i - 16];
+      words[i]      = rotl32(temp, 1);
+    }
+
+#pragma unroll
```

Comment:
I'm going to guess the compiler will be grumpy about unrolling 80 iterations of this loop. You should check if unrolling helps or hurts performance here.